### PR TITLE
feat: dependent features

### PR DESCRIPTION
**Pull request by kwasniew**
## About the changes

Dependent features implementation based on spec: https://github.com/Unleash/client-specification/pull/63

Extra checks that may also be covered by other SDKs (listing them since this is our reference implementation):
* metrics are not called on dependent features
* impression events are called on dependent features for easier debugging
* warning events for missing dependencies are reported once

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -55,8 +55,37 @@ export default class UnleashClient extends EventEmitter {
     }
   }
 
+  isParentDependencySatisfied(feature: FeatureInterface | undefined, context: Context) {
```
Comment (sjaanus) [inline src/client.ts:58]: Why undefined?

File: src/client.ts [hunk 2]
```diff
@@ -55,8 +55,37 @@ export default class UnleashClient extends EventEmitter {
...
+    if (!feature?.name) {
```
Comment (sjaanus) [inline src/client.ts:59]: Can feature not have name?

File: src/client.ts [hunk 3]
```diff
@@ -55,8 +55,37 @@ export default class UnleashClient extends EventEmitter {
...
+
+        if (parentFeature?.dependencies?.length) {
+          return false;
+        }
+
+        if (parent.enabled !== false && parent.variants?.length) {
+          return parent.variants.includes(this.getVariant(parent.feature, context).name);
+        }
+
+        return parent.enabled !== false
```
Comment (sjaanus) [inline src/client.ts:75]: Can we make it more readable, and not do negate operator, but instead `===`

File: src/client.ts [hunk 4]
```diff
@@ -55,8 +55,37 @@ export default class UnleashClient extends EventEmitter {
...
+  isParentDependencySatisfied(feature: FeatureInterface | undefined, context: Context) {
+    if (!feature?.name) {
+      return true;
+    }
+
+    if (feature.dependencies && feature.dependencies.length) {
+      return feature.dependencies.every(parent => {
+        const parentFeature = this.repository.getToggle(parent.feature);
+
+        if (parentFeature?.dependencies?.length) {
```
Comment (sjaanus) [inline src/client.ts:67]: Can we make it more readable, Does this mean length exists or length === 0?

File: src/client.ts (hunk 3 above)
Comment (kwasniew) [reply src/client.ts:75]: by default enabled is not provided so it's undefined. enabled: false is a special case for kill-switches in parent flag.This code is explicit about explicit enabled false and everything else.

File: src/client.ts (hunk 2 above)
Comment (kwasniew) [reply src/client.ts:59]: this is what the existing type was telling me so I'm respecting it here

File: src/client.ts (hunk 1 above)
Comment (kwasniew) [reply src/client.ts:58]: existing types indicated so

File: src/client.ts (hunk 4 above)
Comment (kwasniew) [reply src/client.ts:67]: nope, it should also cater for undefined dependencies. Setting === 0 breaks the spec

File: src/client.ts (hunk 3 above)
Comment (kwasniew) [reply src/client.ts]: I reorganized the conditions a bit but the !== false has to stay as it is

File: src/client.ts (hunk 4 above)
Comment (kwasniew) [reply src/client.ts]: move it to precondition check though

File: src/client.ts [hunk 5]
```diff
@@ -55,8 +55,32 @@ export default class UnleashClient extends EventEmitter {
...
+        return this.isEnabled(parent.feature, context, () => false);
+      }
+
+      return !this.isEnabled(parent.feature, context, () => false);
+    });
+  }
+
   isEnabled(name: string, context: Context, fallback: Function): boolean {
     const feature = this.repository.getToggle(name);
+    if(!this.isParentDependencySatisfied(feature, context)) {
```
Comment (sighphyre) [inline src/client.ts:81]: Doesn't this guard clause mean there's no impression data for the toggle if the parent causes it to be disabled?

File: src/client.ts [hunk 6]
```diff
@@ -55,8 +55,32 @@ export default class UnleashClient extends EventEmitter {
     }
   }
 
+  isParentDependencySatisfied(feature: FeatureInterface | undefined, context: Context) {
+    if (!feature?.name || !feature.dependencies?.length) {
```
Comment (sighphyre) [inline src/client.ts:59]: What's with the name check here? Name should always be a non empty string, do I have the dumb or is the same as 

if (!feature?.dependencies?.length) {
  return true;
}

Comment (kwasniew) [reply src/client.ts:59]: still passes the spec. not sure why I did the name check but it's not in the type anymore. I'm hallucinating :D

File: src/client.ts (hunk 5 above)
Comment (kwasniew) [reply src/client.ts]: Done

File: src/client.ts (hunk 6 above)
Comment (kwasniew) [reply src/client.ts]: Removed

File: src/client.ts (hunk 5 above)
Comment (kwasniew) [reply src/client.ts]: added a test to make sure no one makes the same mistake again :)

**Review by sighphyre — COMMENTED** [review]
Glancing over the metrics code, it looks correct but I think it's worth having one or two tests to check that the parent doesn't get metric counts (I'm guessing, don't know if that's what you want). Especially if this is going to be the reference implementation so others know what to do, since the spec doesn't handle that

**Comment by kwasniew** [comment]
@sighphyre addressed all your points :)

File: src/client.ts (hunk 2 above)
Comment (sighphyre) [reply src/client.ts]: Feature's should have a name, but at least in this SDK we allow people to define their own bootstrapper so it's possible people put goop in there. This one is probably mostly legacy but respecting the existing types seems very sane to me

**Review by sighphyre — APPROVED** [review]
Nice! LGTM

**Review by ivarconr — DISMISSED** [review]

**Pull request merged**